pandas/core/algorithms.py

```diff
@@ -142,6 +142,71 @@ def isin(comps, values):
     return f(comps, values)
 
 
+def safe_sort(values, labels=None, na_sentinel=-1, assume_unique=False):
+    """
+    Sort ``values`` and reorder corresponding ``labels``.
+    ``values`` should be unique if ``labels`` is not None.
+    Safe for use with mixed types (int, str), orders ints before strs.
+
+    Parameters
+    ----------
+    values : ndarray (1-d)
+        Sequence; must be unique if ``labels`` is not None.
+    labels : ndarray (1-d)
+        Indices to ``values``
+    na_sentinel : int, default -1
+        Value in ``labels`` to mark "not found".
+        Ignored when ``labels`` is None.
+    assume_unique : bool, default False
+        When True, doesn't check for uniqness of ``values``.
+        Ignored when ``labels`` is None.
+
+    Returns
+    -------
+    ordered : ndarray
+        Sorted ``values``
+    new_labels : ndarray
+        Reordered ``labels``; returned when ``labels`` is not None
+    """
+    def sort_mixed(values):
+        # order ints before strings, safe in py3
+        str_pos = np.array([isinstance(x, string_types) for x in values],
+                           dtype=bool)
+        nums = np.sort(values[~str_pos])
+        strs = np.sort(values[str_pos])
+        return com._ensure_object(np.concatenate([nums, strs]))
+
+    sorter = None
+    try:
```

Comment:
I think you need something like

```
if not is_list_like(values):
    raise ValueError(.....must be a list-like)
values = np.array(values, copy=False)
```

same for labels, these are cheap checks that you have ndarrays passed (e.g. a Series would work fine here).

alternatively, you can simply check if its an ndarray and raise otherwise (this fits your guarantees as well).